### PR TITLE
feat(medusa): Convert FulfillmentService to TypeScript

### PR DESCRIPTION
**Pull request by pKorsholm**
Fixes CORE-348

### CODE REVIEW

File: packages/medusa-interfaces/src/fulfillment-service.js [hunk 1]
```diff
@@ -24,20 +24,23 @@ class BaseFulfillmentService extends BaseService {
    * to create shipping options in Medusa that can be chosen between by the
    * customer.
    */
-  getFulfillmentOptions() {}
+  getFulfillmentOptions() {
+    throw Error("validateFulfillmentData must be overridden by the child class")
```
Comment (adrien2p) [inline packages/medusa-interfaces/src/fulfillment-service.js:28]: todo: wrong method name in the error

File: packages/medusa-interfaces/src/fulfillment-service.js [hunk 2]
```diff
@@ -56,11 +59,15 @@ class BaseFulfillmentService extends BaseService {
...
    * Used to calculate a price for a given shipping option.
    */
-  calculatePrice(data, cart) {
+  calculatePrice(optionData, data, cart) {
     throw Error("calculatePrice must be overridden by the child class")
   }
 
-  createFulfillment() {
+  createFulfillment(data, items, order, fulfillment) {
+    throw Error("createOrder must be overridden by the child class")
```
Comment (adrien2p) [inline packages/medusa-interfaces/src/fulfillment-service.js:67]: todo: wrong method name in the error

File: packages/medusa-interfaces/src/fulfillment-service.js [hunk 3]
```diff
@@ -56,11 +59,15 @@ class BaseFulfillmentService extends BaseService {
...
+  }
+
+  cancelFulfillment(fulfillment) {
     throw Error("createOrder must be overridden by the child class")
```
Comment (adrien2p) [inline packages/medusa-interfaces/src/fulfillment-service.js:71]: todo: wrong method name in the error

File: packages/medusa/src/services/fulfillment-provider.ts [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+    const { manager } = container
+
+    this.container_ = container
+    this.manager_ = manager
+  }
+
+  async registerInstalledProviders(providers: string[]): Promise<void> {
+    return await this.atomicPhase_(async (manager) => {
+      const { fulfillmentProviderRepository } = this.container_
```
Comment (adrien2p) [inline packages/medusa/src/services/fulfillment-provider.ts:43]: todo: this should be assigned on the correct property of the class and assigned in the constructor

File: packages/medusa/src/services/fulfillment-provider.ts [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+      for (const p of providers) {
+        const n = fulfillmentProviderRepo.create({ id: p, is_installed: true })
+        await fulfillmentProviderRepo.save(n)
+      }
+    })
+  }
+
+  async list(): Promise<FulfillmentProvider[]> {
+    const { manager, fulfillmentProviderRepository } = this.container_
```
Comment (adrien2p) [inline packages/medusa/src/services/fulfillment-provider.ts:57]: todo: this should be assigned on the correct property of the class and assigned in the constructor

File: packages/medusa/src/services/fulfillment-provider.ts [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    )
+  }
+
+  /**
+   * @param {string} provider_id - the provider id
+   * @return {BaseFulfillmentService} the payment fulfillment provider
+   */
+  retrieveProvider(provider_id: string): BaseFulfillmentService {
+    try {
+      return this.container_[`fp_${provider_id}`]
```
Comment (adrien2p) [inline packages/medusa/src/services/fulfillment-provider.ts:87]: todo: this should be the type of the container to restrict the access. You can look here https://github.com/medusajs/medusa/blob/master/packages/medusa/src/services/notification.ts#L20

File: packages/medusa/src/services/fulfillment-provider.ts [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    const provider = this.retrieveProvider(method.shipping_option.provider_id)
+    return provider.createFulfillment(
+      method.data,
+      items,
+      order,
+      fulfillment
+    ) as unknown as Record<string, unknown>
+  }
+
+  async canCalculate(option): Promise<boolean> {
```
Comment (adrien2p) [inline packages/medusa/src/services/fulfillment-provider.ts:111]: todo: Missing typings

File: packages/medusa/src/services/fulfillment-provider.ts [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    const provider = this.retrieveProvider(fulfillment.provider_id)
+    return provider.cancelFulfillment(
+      fulfillment.data
+    ) as unknown as Fulfillment
+  }
+
+  async calculatePrice(
+    option: ShippingOption,
+    data: Record<string, unknown>,
+    cart: Order | Cart | undefined
```
Comment (adrien2p) [inline packages/medusa/src/services/fulfillment-provider.ts:139]: suggestion: `cart?: Order | Cart`

File: packages/medusa/src/services/fulfillment-provider.ts [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+    const provider = this.retrieveProvider(option.provider_id)
+    return provider.createReturn(returnOrder) as unknown as Record<
+      string,
+      unknown
+    >
+  }
+
+  /**
+   * Fetches documents from the fulfillment provider
+   * @param {string} providerId - the id of the provider
```
Comment (adrien2p) [inline packages/medusa/src/services/fulfillment-provider.ts:161]: suggestion: Remove the comments typings in favour of the one in the signature

**Review by adrien2p — COMMENTED** [review]
Small comments :)

File: packages/medusa/src/services/fulfillment-provider.ts [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+  Return,
+  ShippingMethod,
+  ShippingOption,
+} from "../models"
+import { FulfillmentProviderRepository } from "../repositories/fulfillment-provider"
+import { CreateFulfillmentOrder } from "../types/fulfillment"
+import { FulfillmentOptions } from "../types/fulfillment-provider"
+import { MedusaContainer } from "../types/global"
+
+type NotificationProviderKey = `fp_${string}`
```
Comment (olivermrbl) [inline packages/medusa/src/services/fulfillment-provider.ts:20]: todo: Rename
```suggestion
type FulfillmentProviderKey = `fp_${string}`
```

File: packages/medusa/src/services/fulfillment-provider.ts [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+
+type FulfillmentProviderContainer = MedusaContainer & {
+  fulfillmentProviderRepository: typeof FulfillmentProviderRepository
+  manager: EntityManager
+} & {
+  [key in `${NotificationProviderKey}`]: BaseFulfillmentService
```
Comment (olivermrbl) [inline packages/medusa/src/services/fulfillment-provider.ts:26]:
```suggestion
  [key in `${FulfillmentProviderKey}`]: BaseFulfillmentService
```

File: packages/medusa/src/services/fulfillment-provider.ts [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+  async list(): Promise<FulfillmentProvider[]> {
+    const fpRepo = this.manager_.getCustomRepository(
+      this.fulfillmentProviderRepository_
+    )
+
+    return await fpRepo.find({})
+  }
+
+  async listFulfillmentOptions(
+    provider_ids: string[]
```
Comment (olivermrbl) [inline packages/medusa/src/services/fulfillment-provider.ts:73]: todo: Should be camelCased
```suggestion
    providerIds: string[]
```

File: packages/medusa/src/services/fulfillment-provider.ts [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+  ): Promise<FulfillmentOptions[]> {
+    return await Promise.all(
+      provider_ids.map(async (p) => {
```
Comment (olivermrbl) [inline packages/medusa/src/services/fulfillment-provider.ts:76]:
```suggestion
      providerIds.map(async (p) => {
```

File: packages/medusa/src/services/fulfillment-provider.ts [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+        }
+      })
+    )
+  }
+
+  /**
+   * @param provider_id - the provider id
+   * @return the payment fulfillment provider
+   */
+  retrieveProvider(provider_id: string): BaseFulfillmentService {
```
Comment (olivermrbl) [inline packages/medusa/src/services/fulfillment-provider.ts:94]: todo: Should be camelCased
```suggestion
  retrieveProvider(providerId: string): BaseFulfillmentService {
```

File: packages/medusa/src/services/fulfillment-provider.ts [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+    try {
+      return this.container_[`fp_${provider_id}`]
```
Comment (olivermrbl) [inline packages/medusa/src/services/fulfillment-provider.ts:96]:
```suggestion
      return this.container_[`fp_${providerId}`]
```

File: packages/medusa/src/services/fulfillment-provider.ts [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+    } catch (err) {
+      throw new MedusaError(
+        MedusaError.Types.NOT_FOUND,
+        `Could not find a fulfillment provider with id: ${provider_id}`
```
Comment (olivermrbl) [inline packages/medusa/src/services/fulfillment-provider.ts:100]:
```suggestion
        `Could not find a fulfillment provider with id: ${providerId}`
```

**Review by olivermrbl — CHANGES_REQUESTED** [review]

**Review by olivermrbl — APPROVED** [review]
LGTM!

**Pull request merged**